lib/models/patch/file-patch.js

```diff
@@ -0,0 +1,283 @@
+    const lastHunk = hunks[hunks.length - 1];
+    if (!lastHunk) {
+      return [];
+    }
+
+    const range = lastHunk.getNoNewlineRange();
+    if (!range) {
+      return [];
+    }
+
+    return [range];
+  }
+
+  adoptBufferFrom(prevFilePatch) {
+    this.getPatch().adoptBufferFrom(prevFilePatch.getPatch());
+  }
+
+  didChangeExecutableMode() {
+    if (!this.oldFile.isPresent() || !this.newFile.isPresent()) {
+      return false;
+    }
+
+    return this.oldFile.isExecutable() && !this.newFile.isExecutable() ||
+      !this.oldFile.isExecutable() && this.newFile.isExecutable();
+  }
+
+  hasSymlink() {
+    return Boolean(this.getOldFile().getSymlink() || this.getNewFile().getSymlink());
+  }
+
+  hasTypechange() {
+    if (!this.oldFile.isPresent() || !this.newFile.isPresent()) {
+      return false;
+    }
+
+    return this.oldFile.isSymlink() && !this.newFile.isSymlink() ||
+      !this.oldFile.isSymlink() && this.newFile.isSymlink();
+  }
+
+  getPath() {
+    return this.getOldPath() || this.getNewPath();
+  }
+
+  getStatus() {
+    return this.getPatch().getStatus();
+  }
+
+  getHunks() {
+    return this.getPatch().getHunks();
+  }
+
+  clone(opts = {}) {
+    return new this.constructor(
+      opts.oldFile !== undefined ? opts.oldFile : this.oldFile,
+      opts.newFile !== undefined ? opts.newFile : this.newFile,
+      opts.patch !== undefined ? opts.patch : this.patch,
+    );
+  }
+
+  getStagePatchForLines(selectedLineSet) {
+    if (this.patch.getChangedLineCount() === selectedLineSet.size) {
+      if (this.hasTypechange() && this.getStatus() === 'deleted') {
+        // handle special case when symlink is created where a file was deleted. In order to stage the file deletion,
+        // we must ensure that the created file patch has no new file
+        return this.clone({newFile: nullFile});
+      } else {
+        return this;
+      }
+    } else {
+      const patch = this.patch.getStagePatchForLines(selectedLineSet);
+      if (this.getStatus() === 'deleted') {
+        // Populate newFile
+        return this.clone({newFile: this.getOldFile(), patch});
+      } else {
+        return this.clone({patch});
+      }
+    }
+  }
+
+  getStagePatchForHunk(selectedHunk) {
+    return this.getStagePatchForLines(new Set(selectedHunk.getBufferRows()));
+  }
+
+  getUnstagePatchForLines(selectedLineSet) {
+    if (this.patch.getChangedLineCount() === selectedLineSet.size) {
+      const patch = this.patch.getFullUnstagedPatch();
+      if (this.hasTypechange() && this.getStatus() === 'added') {
+        // handle special case when a file was created after a symlink was deleted.
+        // In order to unstage the file creation, we must ensure that the unstage patch has no new file,
+        // so when the patch is applied to the index, there file will be removed from the index.
+        return this.clone({oldFile: nullFile, patch});
+      } else {
+        return this.clone({patch});
```

Comment:
I believe there's a bug here too.

Repro steps - 
1. Create new file with one line of text
1. Stage file
1. Select line and click "Unstage Selected Line" button

Diff (output of `this.clone({patch}).toString()`):
```
"diff --git a/foo.txt b/foo.txt
deleted file mode null
--- /dev/null
+++ b/foo.txt
@@ -1,2 +1,0 @@
-asdf
-
"
```

Error thrown:
```
git-shell-out-strategy.js? [sm]:32 Uncaught (in promise) GitError: git apply --cached - in /Users/kuychaco/src/test-repo exited with code 128
stdout: 
stderr: error: invalid mode on line 2: null
--- /dev/null
+++ b/foo.txt
@@ -1,2 +1,0 @@
-asdf
-


    at new GitError (/Users/kuychaco/github/github/lib/git-shell-out-strategy.js:34:18)
    at Promise (/Users/kuychaco/github/github/lib/git-shell-out-strategy.js:317:23)
    at <anonymous>
```